### PR TITLE
Semi supervised approximator with self-consistency

### PR DESCRIPTION
**Pull request by Kucharssim**
First implementation of the self consistency loss in the new version of bayesflow. It still has some rough edges and we need to test it more properly, but it should be fine for the basic use cases outlined in the `SelfConsistency.ipynb` example notebook. For now we only provide it in the experimental module, but the changes also touch on some other things, mainly to make it possible to run adapter, draw posterior samples, and evaluate log prob using keras operations only (so that we can sample and evaluate log prob during training.). I am also not sure about some of the naming, but first I wanted to get your impression.

### CODE REVIEW

**Comment by Kucharssim** [comment]
the tests are not looking great :D I will check that out

**Comment by Kucharssim** [comment]
I need to add some tests yet to improve coverage, but can you let me know what you think of this so far, @paul-buerkner, @stefanradev93? Thank you!

I recommend looking into the jupyter notebook under `examples/experimental/SelfConsistency.ipynb`

**Comment by paul-buerkner** [comment]
Thanks! I made some edits to the notebook. Here are some more comments that perhaps you can address:

- Provide links to our papers on this method so people can read more about it. 
- Explain the high-level idea of the SC loss earlier.
- Explain why we call it "SemiSupervised".
- Results of the posterior-likelihood approximator are not discussed. The notebook just ends with training. Does it make sense to add a bit more so the notebook doesn't end up that sudden?
- rename the notebook to Self_Consistency (with _) for naming consistency

**Review by paul-buerkner — CHANGES_REQUESTED** [review]
See my comments on the notebook above.

**Review by paul-buerkner — APPROVED** [review]
Thank you! Notebook looks good to me now.

**Comment by paul-buerkner** [comment]
@stefanradev93 can you please review too so we can then merge?

File: bayesflow/experimental/self_consistency/semi_supervised_approximator.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            "composite_metrics": self.composite_metrics,
+        }
+        return base_config | serialize(config)
+
+    def build(self, data_shapes: Mapping[str, tuple[int] | Mapping[str, Mapping]]):
+        for name, appr in self.approximators.items():
+            if hasattr(appr, "built") and not appr.built:
+                appr.build(data_shapes[name])
+
+        self.built = True
```
Comment (stefanradev93) [inline bayesflow/experimental/self_consistency/semi_supervised_approximator.py]: Not needed.

File: bayesflow/experimental/self_consistency/semi_supervised_approximator.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    def compute_metrics(self, stage: str = "training", **data: Mapping[str, Tensor]) -> dict[str, Tensor]:
+        """Compute the total loss and all sub-metrics for one training step.
+
+        Parameters
+        ----------
+        **data : Mapping[str, Tensor]
+            Named list of input datasets.
+        stage : str, optional
+            Current training stage (e.g., "training", "validation", "inference"). Controls
+            the behavior of standardization and some metric computations (default is "training").
```
Comment (stefanradev93) [inline bayesflow/experimental/self_consistency/semi_supervised_approximator.py]: Not the correct order of args.

File: bayesflow/experimental/self_consistency/semi_supervised_approximator.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        """
+        if hasattr(self.approximators[approximator], "sample"):
+            samples = self.approximators[approximator].sample(**kwargs)
+        else:
+            raise AttributeError(f"approximator `{approximator}` does not have a `sample` method.")
+
+        summary_outputs = samples.pop("_summaries", None)
+
+        if self.adapter is not None:
+            samples = self.adapter(samples, inverse=True)
```
Comment (stefanradev93) [inline bayesflow/experimental/self_consistency/semi_supervised_approximator.py]: Adapter cannot be `None` given the existing signature.

File: bayesflow/experimental/self_consistency/semi_supervised_approximator.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+        if summary_outputs is not None:
+            samples["_summaries"] = summary_outputs
```
Comment (stefanradev93) [inline bayesflow/experimental/self_consistency/semi_supervised_approximator.py]: This, together with the above `pop` is unnecessary. I would just return the result from `sample`.

File: bayesflow/experimental/self_consistency/semi_supervised_approximator.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        self.build(keras.tree.map_structure(keras.ops.shape, adapted_data))
+
+    def _adapt_data(self, data: Mapping[str, Any]) -> Mapping[str, Any]:
+        adapted_data = {}
+
+        for name, appr in self.approximators.items():
+            if hasattr(appr, "adapter"):
+                adapted_data[name] = appr.adapter(data)
+            else:
+                adapted_data[name] = data
```
Comment (stefanradev93) [inline bayesflow/experimental/self_consistency/semi_supervised_approximator.py]: An approximator always has an adapter, no?

File: bayesflow/approximators/approximator.py [hunk 6]
```diff
@@ -153,9 +153,9 @@ def _prepare_conditions(
         summary_outputs : Tensor or None
             Raw summary network outputs, or ``None`` if no summary network.
         """
-
         if data is not None:
-            adapted = self.adapter(data, strict=False, **kwargs)
+            data = keras.tree.map_structure(keras.ops.convert_to_tensor, data)
+            adapted = self.adapter(data, strict=False, keras=True, **kwargs)
             adapted = keras.tree.map_structure(keras.ops.convert_to_tensor, adapted)
```
Comment (stefanradev93) [inline bayesflow/approximators/approximator.py]: Adds one unnecessary call for non-keras adapters.

File: bayesflow/approximators/helpers/conditions.py [hunk 7]
```diff
@@ -94,24 +94,31 @@ def resolve(
...
-                    }
-
-                    batch_outputs = summary_network(
-                        batch_variables, **filter_kwargs(batch_kwargs, summary_network.call)
-                    )
-                    batches.append(batch_outputs)
 
-                summary_outputs = tree_concatenate(batches, axis=0)
+                if num_conditions is None:
+                    # Graph mode: shapes are symbolic, cannot iterate over batches
```
Comment (stefanradev93) [inline bayesflow/approximators/helpers/conditions.py]: Not sure I understand this comment. Can you please clarify?

File: bayesflow/adapters/adapter.py
Comment (stefanradev93) [inline bayesflow/adapters/adapter.py:41]: General remark. I think the fact that we can support keras ops in the adapter is great and I commend the change. However wouldn't it be easier to leave it to the caller to simply create the `Adapter` with a flag `use_keras` (or similiar) such that the adapter then internally knows which op to use.  From my POV, this would remove the need to shuffle a `keras` kwarg throughout the entire pipeline, and it is then very easy to reconfigure the whole adapter by doing something like `adapter.use_keras=False` if ever needed. @paul-buerkner @Kucharssim What do you think?

File: pyproject.toml [hunk 8]
```diff
@@ -61,6 +61,7 @@ test = [
   "pytest-cov>=7.0.0",
   "pytest-rerunfailures>=16.1",
   "pytest-xdist>=3.8.0",
+  "tensorflow-probability>=0.25.0"
```
Comment (stefanradev93) [inline pyproject.toml:65]: Not sure I like having tf prob as a dependency. If anything, it should be an optional dependency. By the same token, we can argue for `numpyro` or similar as a dep.

**Review by stefanradev93 — CHANGES_REQUESTED** [review]

File: bayesflow/adapters/adapter.py
Comment (Kucharssim) [reply bayesflow/adapters/adapter.py:41]: Yes, that was the other way to do it. I am generally fine implementing it either way!

File: pyproject.toml (hunk 8 above)
Comment (Kucharssim) [reply pyproject.toml:65]: Is it used only for testing, and it is specified among the `test` dependencies. Are these not optional?

I am happy to rewrite the tests without relying on this, but I thought it would be nice to cover the use-case of users specifying prior/likelihood using tensorflow distributions, as it makes using self-consistency much easier...

**Comment by paul-buerkner** [comment]
About adapter design, both is fine by me. If you two agree the other one is
nicer, feel free to change it

Simon Kucharsky ***@***.***> schrieb am Mi. 8. Juli 2026 um
21:59:

File: pyproject.toml (hunk 8 above)
Comment (stefanradev93) [reply pyproject.toml:65]: Missed that on the github view, happy to keep it as is.

File: bayesflow/approximators/helpers/conditions.py (hunk 7 above)
Comment (Kucharssim) [reply bayesflow/approximators/helpers/conditions.py]: I've expanded on the comment, is it now clearer?

File: bayesflow/adapters/transforms/to_dict.py [hunk 9]
```diff
@@ -28,12 +29,12 @@ def forward(self, data, **kwargs) -> dict[str, np.ndarray]:
                 if value.dtype == "category":
                     value = pd.get_dummies(value)
 
-                value = np.asarray(value).astype("float32", copy=False)
+                value = ops.convert_to_tensor(value, dtype="float32")
```
Comment (stefanradev93) [inline bayesflow/adapters/transforms/to_dict.py]: Shouldn't we use floatx here?

File: bayesflow/adapters/adapter.py [hunk 10]
```diff
@@ -100,7 +103,7 @@ def get_config(self) -> dict:
         return serialize(config)
 
     def forward(
-        self, data: dict[str, Any], *, log_det_jac: bool = False, **kwargs
```
Comment (stefanradev93) [inline bayesflow/adapters/adapter.py:103]: `Any` from `typing` is the recommended hint, afaik.

File: bayesflow/adapters/adapter.py [hunk 11]
```diff
@@ -60,14 +59,18 @@ class Adapter(MutableSequence[Transform]):
     ----------
     transforms : Sequence[Transform], optional
         The sequence of transforms to execute.
+    device_name: str, optional
+        Passed to keras.device to specify on which device to apply the adapter.
     """
 
-    def __init__(self, transforms: Sequence[Transform] | None = None):
+    def __init__(self, transforms: Sequence[Transform] | None = None, device_name: str = "cpu"):
```
Comment (stefanradev93) [inline bayesflow/adapters/adapter.py]: Should we just go with `device` as in canonical in torch?

File: bayesflow/approximators/continuous_approximator.py [hunk 12]
```diff
@@ -222,6 +222,7 @@ def sample(
         sample_shape: Literal["infer"] | Tuple[int] | int = "infer",
         return_summaries: bool = False,
         seed: int | keras.random.SeedGenerator | None = None,
+        numpy: bool = True,
```
Comment (stefanradev93) [inline bayesflow/approximators/continuous_approximator.py]: Missing in docstring and can be dangerously mistaken for the library. Perhaps `to_numpy` / `as_numpy`?

File: bayesflow/approximators/continuous_approximator.py [hunk 13]
```diff
@@ -299,7 +302,7 @@ def sample(
 
         return samples
 
-    def log_prob(self, data: Mapping[str, np.ndarray], **kwargs) -> np.ndarray:
+    def log_prob(self, data: Mapping[str, np.ndarray], numpy: bool = True, **kwargs) -> np.ndarray:
```
Comment (stefanradev93) [inline bayesflow/approximators/continuous_approximator.py]: Same as above.

File: bayesflow/experimental/self_consistency/semi_supervised_approximator.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+        Generates samples from one of the approximator (specified by its name) given input conditions.
+        **kwargs are passed to the approximator.sample method.
+        """
+        if hasattr(self.approximators[approximator], "sample"):
+            samples = self.approximators[approximator].sample(**kwargs)
+        else:
+            raise AttributeError(f"approximator `{approximator}` does not have a `sample` method.")
+
+        samples = self.adapter(samples, inverse=True)
+        if numpy:
```
Comment (stefanradev93) [inline bayesflow/experimental/self_consistency/semi_supervised_approximator.py]: Same comment regarding naming.

File: bayesflow/utils/tensor_utils.py [hunk 15]
```diff
@@ -37,6 +37,14 @@ def concatenate_valid_shapes(tensor_shapes: Sequence[Shape | None], axis: int =
     return output_shape
 
 
+def repeat_valid(x: Tensor, repeats: int, axis: int = 0) -> Tensor | None:
```
Comment (stefanradev93) [inline bayesflow/utils/tensor_utils.py]: type hint should be `Tenor | None`

File: tests/test_adapters/test_adapters.py [hunk 16]
```diff
@@ -61,14 +62,16 @@ def test_constrain():
 
     ad = (
         Adapter()
+        .to_array()
+        .convert_dtype("float64", "float32")
         .constrain("x_lower_cont", lower=0)
         .constrain("x_upper_cont", upper=0)
         .constrain("x_both_cont", lower=0, upper=1)
-        .constrain("x_lower_disc1", lower=0, inclusive="lower")
+        .constrain("x_lower_disc1", lower=0, inclusive="lower", epsilon=1e-6)
```
Comment (stefanradev93) [inline tests/test_adapters/test_adapters.py]: Why is this not default if it's important for stability?

File: bayesflow/adapters/transforms/random_subsample.py [hunk 17]
```diff
@@ -26,20 +29,19 @@ def __init__(
...
-    def forward(self, data: np.ndarray, **kwargs) -> np.ndarray:
-        max_sample_size = data.shape[self.axis]
+    def forward(self, data: Tensor, **kwargs) -> Tensor:
+        max_sample_size = ops.shape(data)[self.axis]
 
         if isinstance(self.sample_size, int):
             sample_size = self.sample_size
         else:
-            sample_size = np.round(self.sample_size * max_sample_size)
+            sample_size = int(ops.round(self.sample_size * max_sample_size))
```
Comment (stefanradev93) [inline bayesflow/adapters/transforms/random_subsample.py:38]: Why does this even need an `ops` call?

**Review by stefanradev93 — CHANGES_REQUESTED** [review]
Great work, Simon! I think we are almost there. Did you do some runtime benchmarks with the new (CPU / GPU) adapter vs. the old numpy one?

File: bayesflow/adapters/adapter.py (hunk 11 above)
Comment (Kucharssim) [reply bayesflow/adapters/adapter.py]: I am fine with that, but since we are only using it to pass it forward to keras.device, I figured I can keep consistent with that

File: bayesflow/adapters/transforms/random_subsample.py (hunk 17 above)
Comment (Kucharssim) [reply bayesflow/adapters/transforms/random_subsample.py:38]: We can keep the numpy call, I just removed it so that we can remove the numpy import. I don't mind either way...

File: bayesflow/adapters/transforms/to_dict.py (hunk 9 above)
Comment (Kucharssim) [reply bayesflow/adapters/transforms/to_dict.py]: Yes, thanks!

File: tests/test_adapters/test_adapters.py (hunk 16 above)
Comment (Kucharssim) [reply tests/test_adapters/test_adapters.py]: Good point, will change that

File: bayesflow/adapters/adapter.py (hunk 10 above)
Comment (Kucharssim) [reply bayesflow/adapters/adapter.py:103]: Ok, changed throughout the adapter

**Comment by Kucharssim** [comment]
> Did you do some runtime benchmarks with the new (CPU / GPU) adapter vs. the old numpy one?

I've only run benchmarks comparing CPU vs GPU on the new adapter. I will compare to the old numpy as well
